Low level failure when calculations get large on Mono

### Description
Cells sometimes silently fail to execute variable assignments. 
Errors are logged in shell.log

### Repro steps
1. Unzip and run the attached notebook file. 

### Expected behavior
All the code in this notebook should run without error. 

### Actual behavior
Code for the first example fails to return its results to the IfSharp notebook. 

### Known workarounds
* The code in the cell still executes and it is possible to save the intended values to a system file
while the variable in the notebook interface is not created/updated.  
* Breaking up code into small one-liners in individual cells
* Sometimes it helps to pipe instead of declaring intermediate variables.
* Running smaller examples. 

### Related information 
This problem has existed in IfSharp at least version 1.0 on both OSx and Linux. 
Problem code works fine in the fsharpi.exe environment. 
Problem code runs fine on Windows with .net (not tested with Mono on Windows).


* Operating system : Ubuntu 16.04, OSx 10.11.6 
* Branch master
* Mono Version 4.8.1

[problem book.ipynb.zip](https://github.com/fsprojects/IfSharp/files/929285/problem.book.ipynb.zip)

Thanks @dgrechka that substantially narrows down the issue!

I'm able to reproduce on the Linux System System for Linux (https://www.mono-project.com/download/stable/#download-lin):

```
sudo apt-key adv --keyserver hkp://keyserver.ubuntu.com:80 --recv-keys 3FA7E0328081BFF6A14DA29AA6A19B38D3D831EF
sudo apt install apt-transport-https
echo "deb https://download.mono-project.com/repo/ubuntu stable-xenial main" | sudo tee /etc/apt/sources.list.d/mono-official-stable.list
sudo apt update

sudo apt install mono-devel

sudo apt-get install fsharp
```

and results in:

```
./run_repro.sh
System info
Linux WINDOWS-XXXXXXX 4.4.0-17134-Microsoft #81-Microsoft Sun May 20 01:14:00 PST 2018 x86_64 x86_64 x86_64 GNU/Linux
No LSB modules are available.
Distributor ID: Ubuntu
Description:    Ubuntu 16.04.3 LTS
Release:        16.04
Codename:       xenial
Mono JIT compiler version 5.12.0.226 (tarball Thu May  3 09:48:32 UTC 2018)
Copyright (C) 2002-2014 Novell, Inc, Xamarin Inc and Contributors. www.mono-project.com
        TLS:           __thread
        SIGSEGV:       altstack
        Notifications: epoll
        Architecture:  amd64
        Disabled:      none
        Misc:          softdebug
        Interpreter:   yes
        LLVM:          supported, not enabled.
        GC:            sgen (concurrent by default)
Clearing outputs
Building executable with fsharpc...
Microsoft (R) F# Compiler version 4.1
Copyright (c) Microsoft Corporation. All Rights Reserved.
Gathering dependencies...
Triggering issue by running executable...

Unhandled Exception:
StackOverflowException
[ERROR] FATAL UNHANDLED EXCEPTION: System.StackOverflowException: The requested operation caused a stack overflow.
```
(full log attached)
[repo_WSL.txt](https://github.com/fsprojects/IfSharp/files/2053516/repo_WSL.txt)

The current direct dependencies are:

```
FSharp.Compiler.Service (13.0)
FSharp.Compiler.Tools (4.1.29)
```

I upgraded each to their latest:

```
nuget FSharp.Compiler.Service ~> 22.0.3
nuget FSharp.Compiler.Tools ~> 10.0.2
```

Some details are changed but the overall result is the same:

```
[ERROR] FATAL UNHANDLED EXCEPTION: System.StackOverflowException: The requested operation caused a stack overflow.
  at Microsoft.FSharp.Compiler.Layout.addL@388-4[b,a,c] (Microsoft.FSharp.Compiler.Layout+LayoutRenderer`2[a,b] rr, a z, System.Int32 pos, System.Int32 i, Internal.Utilities.StructuredFormat.Layout layout, Microsoft.FSharp.Core.FSharpFunc`2[T,TResult] k) <0x7fb7f150 + 0x00007> in <5abc127aa80f33aea74503837a12bc5a>:0
  at Microsoft.FSharp.Compiler.Layout+addL@388-5T[a,b,c].Invoke (a z, System.Int32 pos, System.Int32 i, Internal.Utilities.StructuredFormat.Layout layout, Microsoft.FSharp.Core.FSharpFunc`2[T,TResult] k) [0x00007] in <5abc127aa80f33aea74503837a12bc5a>:0
  at Microsoft.FSharp.Core.OptimizedClosures+Invoke@3307-3[T2,T3,T4,T5,TResult,T1].Invoke (T2 u, T3 v, T4 w, T5 x) [0x00001] in <5893d081904cf4daa745038381d09358>:0
  at Microsoft.FSharp.Core.OptimizedClosures+Invoke@3301-2[T2,T3,T4,TResult,T1].Invoke (T2 u, T3 v, T4 w) [0x00001] in <5893d081904cf4daa745038381d09358>:0
  at Microsoft.FSharp.Core.OptimizedClosures+Invoke@3266-1[T2,T3,TResult,T1].Invoke (T2 u, T3 v) [0x00001] in <5893d081904cf4daa745038381d09358>:0
  at Microsoft.FSharp.Core.OptimizedClosures+Invoke@3253[T2,TResult,T1].Invoke (T2 u) [0x00001] in <5893d081904cf4daa745038381d09358>:0
```

--- a/src/IfSharp.Kernel/Evaluation.fs
+++ b/src/IfSharp.Kernel/Evaluation.fs
@@ -42,6 +42,14 @@ module Evaluation =
         let printStream = new StringWriter(sbPrint)
     
         let fsiObj = Microsoft.FSharp.Compiler.Interactive.Settings.fsi
+        // The following is a workaround for IfSharp github issue #143
+        // Mono fails to handle tailcall in Fsi printing code, thus constraining the length of print on Mono
+        if Type.GetType("Mono.Runtime") <> null then
+            // Default PrintLength value of 100 triggeres the issue when printing the array of size 12x100
+            // Value of 50 "postpones" the issue. E.g. Issue is triggered on printing larger arrays of size 50x100
+            // Value of 10 seems to work, arrays 1000x1000 are printed without the error, although trancated with elipsis as expected
+            // After the value is set to 10, effectivly the sequences are truncated to 100 elements during printing . Maybe F# interactive issue
+            fsiObj.PrintLength <- 10            
         let fsiConfig = FsiEvaluationSession.GetDefaultConfiguration(fsiObj, false)
         let args = [|"--noninteractive"; "--define:HAS_FSI_ADDHTMLPRINTER" |]
         let fsiSession = FsiEvaluationSession.Create(fsiConfig, args, inStream, outStream, errStream)

